include/triton/core/tritonbackend.h

```diff
@@ -1356,6 +1358,154 @@ TRITONBACKEND_ModelInstanceReportStatistics(
+TRITONBACKEND_ModelInstanceResponseStatisticsNew(
+    TRITONBACKEND_ModelInstanceResponseStatistics** response_statistics);
+
+/// Delete an inference response statistics object.
+///
+/// The caller retains ownership to the objects set on the deleted response
+/// statistics object and must free them separately.
+///
+/// \param response_statistics The response statistics object to be deleted.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONBACKEND_DECLSPEC TRITONSERVER_Error*
+TRITONBACKEND_ModelInstanceResponseStatisticsDelete(
+    TRITONBACKEND_ModelInstanceResponseStatistics* response_statistics);
+
+/// Set model instance to an inference response statistics object.
+///
+/// \param response_statistics The response statistics object.
+/// \param model_instance The model instance.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONBACKEND_DECLSPEC TRITONSERVER_Error*
+TRITONBACKEND_ModelInstanceResponseStatisticsSetModelInstance(
+    TRITONBACKEND_ModelInstanceResponseStatistics* response_statistics,
+    TRITONBACKEND_ModelInstance* model_instance);
+
+/// Set response factory to an inference response statistics object.
+///
+/// \param response_statistics The response statistics object.
+/// \param response_factory The response factory.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONBACKEND_DECLSPEC TRITONSERVER_Error*
+TRITONBACKEND_ModelInstanceResponseStatisticsSetResponseFactory(
+    TRITONBACKEND_ModelInstanceResponseStatistics* response_statistics,
+    TRITONBACKEND_ResponseFactory* response_factory);
+
+/// Set response start time to an inference response statistics object.
+///
+/// All timestamps should be reported in nanonseconds and collected using
+/// std::chrono::steady_clock::now().time_since_epoch() or the equivalent.
+///
+/// For consistency of measurement across different backends, the timestamps
+/// should be collected at the following points during
+/// TRITONBACKEND_ModelInstanceExecute.
+///
+///   TRITONBACKEND_ModelInstanceExecute()
+///     < start of this response >
+///     CAPTURE TIMESTAMP (response_start)
+///     < generate this response >
+///     CAPTURE TIMESTAMP (compute_output_start)
+///     < allocate output buffers and extract output tensors, including copying
+///       the tensors to/from GPU if necessary >
+///     CAPTURE TIMESTAMP (response_end)
+///     < end of this response >
+///     return
+///
+/// \param response_statistics The response statistics object.
+/// \param response_start The response start time.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONBACKEND_DECLSPEC TRITONSERVER_Error*
+TRITONBACKEND_ModelInstanceResponseStatisticsSetResponseStart(
+    TRITONBACKEND_ModelInstanceResponseStatistics* response_statistics,
+    uint64_t response_start);
+
+/// Set compute output start time to an inference response statistics object.
+/// Set this to 0 for reporting an empty response.
```

Comment:
Given that we have an opaque object now, I feel like the correct way to specify empty would some kind of bool or flag. I think this is probably fine as-is, but I wish we had a some example use cases for this feature. In my mind, there could be models that do some compute work, but decide it's not ready for a full response yet. Having some way to specify "work was done" (compute > 0), but also "empty response" may be useful..

It's hard to say without some concrete use case for this feature. What do you think?

Reply:
The response statistics collects 3 timestamps and translates into 2 durations:
```
response start    -->    compute output start    -->    response end
        v                           v                         v
        |       compute infer       |      compute output     |
```
On you example, if there is no compute output start time, which implies the response is empty, then there is no compute output duration which the compute infer duration is from response start to response end, so the graph for empty becomes:
```
response start    -->   response end
        v                     v
        |    compute infer    |
```
I think your example is the generic case where users should use empty response, because some compute work is done but there is no output tensor, so there is no coping of output tensor, which there is no compute output duration. The only purpose for the compute output start timestamp is to break the total duration from response start to end into compute infer and output durations. Since there is no compute output duration for empty response (if and only if), the compute output start timestamp can be repurposed for denoting empty response.

I think it is more natural to use `compute output start = 0` to denote an empty response, because the user should not record the compute output start timestamp if copying output tensors never takes place. Then, the user can simply skip setting compute output start time (or set it to 0 if it fits the workflow better), and it is automatically classified as an empty response, without needing the user to additionally report it is an empty response or not. What do you think?

Reply:
I think `#5` is an example for this:
```
        '5': {
          'compute_infer': {'count': 1, 'ns': 400089434},
          'compute_output': {'count': 0, 'ns': 0},
          'success': {'count': 0, 'ns': 0},
          'fail': {'count': 0, 'ns': 0},
          'empty_response': {'count': 1, 'ns': 400089434}}
        }
```
There is an empty response with compute duration, but not output duration.

Reply:
I think `#3` is an interesting example:
```
        '3': {
          'compute_infer': {'count': 2, 'ns': 800160632},
          'compute_output': {'count': 1, 'ns': 200133237},
          'success': {'count': 0, 'ns': 0},
          'fail': {'count': 1, 'ns': 600214152},
          'empty_response': {'count': 1, 'ns': 400079717}
        },
```
There is a fail and an empty response. Regardless of result, compute infer always take 4 seconds and compute output always take 2 seconds (if any). On the empty response, there is no compute output but compute infer, so the total duration is 4 seconds. On the fail response, there is compute infer and output, so the total duration is 6 seconds. There is only 1 compute output from the fail response, so there is only 1 compute output totaling 2 second. There is a compute infer from fail and empty, so there are 2 compute infer totaling 8 seconds.

Reply:
Looks like "success" here also implies "non-empty" btw.
>          'success': {'count': 0, 'ns': 0},
>          'empty_response': {'count': 1, 'ns': 400079717}

Reply:
Yes, this is expected.

A summary of where the response should be grouped into (i.e. success, fail, empty):

|   | no failure (success) | has failure |
| --- | --- | --- |
| **sent response** | success | fail |
| **no response sent** | empty | fail |